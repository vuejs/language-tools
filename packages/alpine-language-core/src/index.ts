import * as vue from '@volar/vue-language-core';
import * as useHtmlFilePlugin from './plugins/file-html';

export type LanguageServiceHost = vue.LanguageServiceHost;

export function createLanguageContext(host: vue.LanguageServiceHost): vue.LanguageContext {
	return vue.createLanguageContext(host, [useHtmlFilePlugin], ['.html']);
}
