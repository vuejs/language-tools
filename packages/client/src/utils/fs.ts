import * as vscode from 'vscode';

export async function exists(uri: vscode.Uri) {
    try {
        await vscode.workspace.fs.stat(uri);
        return true;
    }
    catch {
        return false;
    }
}
