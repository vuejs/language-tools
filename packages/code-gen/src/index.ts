import type { Mapping, Mode, Range } from '@volar/source-map';

export class CodeGen<T = undefined> {

	private text = '';
	private mappings: Mapping<T>[] = [];

	public getText() {
		return this.text;
	}
	public getMappings(sourceRangeParser?: (data: T, range: Range) => Range): Mapping<T>[] {
		if (!sourceRangeParser) {
			return this.mappings;
		}
		return this.mappings.map(mapping => ({
			...mapping,
			sourceRange: sourceRangeParser(mapping.data, mapping.sourceRange),
			additional: mapping.additional
				? mapping.additional.map(extraMapping => ({
					...extraMapping,
					sourceRange: sourceRangeParser(mapping.data, extraMapping.sourceRange),
				}))
				: undefined,
		}));
	}
	public addCode(str: string, sourceRange: Range, mode: Mode, data: T, extraSourceRanges?: Range[]) {
		const targetRange = this.addText(str);
		this.addMapping2({
			mappedRange: targetRange,
			sourceRange,
			mode,
			data,
			additional: extraSourceRanges ? extraSourceRanges.map(extraSourceRange => ({
				mappedRange: targetRange,
				mode,
				sourceRange: extraSourceRange,
			})) : undefined,
		});
		return targetRange;
	}
	public addMapping(str: string, sourceRange: Range, mode: Mode, data: T) {
		const targetRange = {
			start: this.text.length,
			end: this.text.length + str.length,
		};
		this.addMapping2({ mappedRange: targetRange, sourceRange, mode, data });
		return targetRange;
	}
	public addMapping2(mapping: Mapping<T>) {
		this.mappings.push(mapping);
	}
	public addText(str: string) {
		const range = {
			start: this.text.length,
			end: this.text.length + str.length,
		};
		this.text += str;
		return range;
	}
}

export function margeCodeGen<T extends CodeGen<any>>(a: T, b: T) {
	const aLength = a.getText().length;
	for (const mapping of b.getMappings()) {
		a.addMapping2({
			...mapping,
			mappedRange: {
				start: mapping.mappedRange.start + aLength,
				end: mapping.mappedRange.end + aLength,
			},
			additional: mapping.additional ? mapping.additional.map(mapping_2 => ({
				...mapping_2,
				mappedRange: {
					start: mapping_2.mappedRange.start + aLength,
					end: mapping_2.mappedRange.end + aLength,
				},
			})) : undefined,
		});
	}
	a.addText(b.getText());
}
