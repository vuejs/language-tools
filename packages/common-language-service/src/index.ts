export * from './contextStore';
export * from './plugin';
