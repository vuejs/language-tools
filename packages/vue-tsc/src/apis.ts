import * as shared from '@volar/shared';
import type * as ts from 'typescript/lib/tsserverlibrary';
import type { TypeScriptFeaturesRuntimeContext } from '@volar/vue-typescript';

const lsTypes = ['script', 'template'] as const;

export function register({ typescript: ts, sourceFiles, templateTsLsRaw, scriptTsLsRaw, templateTsHost, scriptTsHost, vueHost }: TypeScriptFeaturesRuntimeContext) {

	return {
		getRootFileNames,
		emit,
		getSyntacticDiagnostics,
		getSemanticDiagnostics,
		getGlobalDiagnostics,
	};

	function getRootFileNames() {
		const set = new Set([
			...getProgram('script').getRootFileNames().filter(fileName => scriptTsHost.fileExists?.(fileName)),
			...getProgram('template').getRootFileNames().filter(fileName => templateTsHost.fileExists?.(fileName)),
		]);
		return [...set.values()];
	}
	function getSyntacticDiagnostics(sourceFile?: ts.SourceFile, cancellationToken?: ts.CancellationToken): readonly ts.DiagnosticWithLocation[] {
		return lsTypes.map(lsType => transformDiagnostics(lsType, getProgram(lsType).getSyntacticDiagnostics(sourceFile, cancellationToken))).flat();
	}
	function getSemanticDiagnostics(sourceFile?: ts.SourceFile, cancellationToken?: ts.CancellationToken): readonly ts.Diagnostic[] {
		return lsTypes.map(lsType => transformDiagnostics(lsType, getProgram(lsType).getSemanticDiagnostics(sourceFile, cancellationToken))).flat();
	}
	function getGlobalDiagnostics(cancellationToken?: ts.CancellationToken): readonly ts.Diagnostic[] {
		return lsTypes.map(lsType => transformDiagnostics(lsType, getProgram(lsType).getGlobalDiagnostics(cancellationToken))).flat();
	}
	function emit(targetSourceFile?: ts.SourceFile, _writeFile?: ts.WriteFileCallback, cancellationToken?: ts.CancellationToken, emitOnlyDtsFiles?: boolean, customTransformers?: ts.CustomTransformers): ts.EmitResult {
		const scriptResult = getProgram('script').emit(targetSourceFile, (vueHost.writeFile ?? ts.sys.writeFile), cancellationToken, emitOnlyDtsFiles, customTransformers);
		const templateResult = getProgram('template').emit(targetSourceFile, undefined, cancellationToken, emitOnlyDtsFiles, customTransformers);
		return {
			emitSkipped: scriptResult.emitSkipped,
			emittedFiles: scriptResult.emittedFiles,
			diagnostics: [
				...transformDiagnostics('script', scriptResult.diagnostics),
				...transformDiagnostics('template', templateResult.diagnostics),
			],
		};
	}
	function getProgram(lsType: 'script' | 'template') {
		const program = (lsType === 'script' ? scriptTsLsRaw : templateTsLsRaw).getProgram();
		if (!program) throw '!program';
		return program;
	}

	// transform
	function transformDiagnostics<T extends ts.Diagnostic | ts.DiagnosticWithLocation | ts.DiagnosticRelatedInformation>(lsType: 'script' | 'template', diagnostics: readonly T[]): T[] {
		const result: T[] = [];
		for (const diagnostic of diagnostics) {
			if (
				diagnostic.file !== undefined
				&& diagnostic.start !== undefined
				&& diagnostic.length !== undefined
			) {
				const fileName = shared.normalizeFileName(diagnostic.file.fileName);
				for (const tsOrVueLoc of sourceFiles.fromTsLocation2(
					lsType,
					shared.fsPathToUri(fileName),
					diagnostic.start,
					diagnostic.start + diagnostic.length,
					data => !!data.capabilities.diagnostic,
				)) {

					if (!vueHost.fileExists?.(shared.uriToFsPath(tsOrVueLoc.uri)))
						continue;

					if (tsOrVueLoc.type === 'source-ts' && lsType !== 'script')
						continue;

					let file = shared.uriToFsPath(tsOrVueLoc.uri) === fileName
						? diagnostic.file
						: undefined;
					if (!file) {

						let docText = tsOrVueLoc.sourceMap?.sourceDocument.getText();

						if (docText === undefined) {
							const snapshot = vueHost.getScriptSnapshot(shared.uriToFsPath(tsOrVueLoc.uri));
							if (snapshot) {
								docText = snapshot.getText(0, snapshot.getLength());
							}
						}

						if (docText !== undefined) {
							file = ts.createSourceFile(shared.uriToFsPath(tsOrVueLoc.uri), docText, tsOrVueLoc.uri.endsWith('.vue') ? ts.ScriptTarget.JSON : ts.ScriptTarget.Latest)
						}
					}
					const newDiagnostic: T = {
						...diagnostic,
						file,
						start: tsOrVueLoc.range.start,
						length: tsOrVueLoc.range.end - tsOrVueLoc.range.start,
					};
					const relatedInformation = (diagnostic as ts.Diagnostic).relatedInformation;
					if (relatedInformation) {
						(newDiagnostic as ts.Diagnostic).relatedInformation = transformDiagnostics(lsType, relatedInformation);
					}
					result.push(newDiagnostic);
				}
			}
			else if (
				diagnostic.file === undefined
			) {
				result.push(diagnostic);
			}
		}
		return result;
	}
}
