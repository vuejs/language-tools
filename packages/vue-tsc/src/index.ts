import type * as ts from 'typescript/lib/tsserverlibrary';
import * as vue from '@vue/language-core';
import * as vueTs from '@vue/typescript';
import { state } from './shared';

export type Hook = (program: _Program) => void;

export type _Program = ts.Program & { __vue: ProgramContext; };

interface ProgramContext {
	projectVersion: number,
	options: ts.CreateProgramOptions,
	languageHost: vue.TypeScriptLanguageHost,
	vueCompilerOptions: Partial<vue.VueCompilerOptions>,
	languageService: ReturnType<typeof vueTs.createLanguageService>,
}

const windowsPathReg = /\\/g;

export function createProgram(options: ts.CreateProgramOptions) {

	if (!options.options.noEmit && !options.options.emitDeclarationOnly)
		throw toThrow('js emit is not supported');

	if (!options.options.noEmit && options.options.noEmitOnError)
		throw toThrow('noEmitOnError is not supported');

	if (options.options.extendedDiagnostics || options.options.generateTrace)
		throw toThrow('--extendedDiagnostics / --generateTrace is not supported, please run `Write Virtual Files` in VSCode to write virtual files and use `--extendedDiagnostics` / `--generateTrace` via tsc instead of vue-tsc to debug.');

	if (!options.host)
		throw toThrow('!options.host');

	const ts = require('typescript') as typeof import('typescript/lib/tsserverlibrary');

	let program = options.oldProgram as _Program | undefined;

	if (state.hook) {
		program = state.hook.program;
		program.__vue.options = options;
	}
	else if (!program) {

		const ctx: ProgramContext = {
			projectVersion: 0,
			options,
			get languageHost() {
				return languageHost;
			},
			get vueCompilerOptions() {
				return vueCompilerOptions;
			},
			get languageService() {
				return vueTsLs;
			},
		};
		const vueCompilerOptions = getVueCompilerOptions();
		const scripts = new Map<string, {
			projectVersion: number,
			modifiedTime: number,
			scriptSnapshot: ts.IScriptSnapshot,
		}>();
		const languageHost: vue.TypeScriptLanguageHost = {
			workspacePath: ctx.options.host!.getCurrentDirectory().replace(windowsPathReg, '/'),
			rootPath: ctx.options.host!.getCurrentDirectory().replace(windowsPathReg, '/'),
			getCompilationSettings: () => ctx.options.options,
			getScriptFileNames: () => {
				return ctx.options.rootNames as string[];
			},
			getScriptSnapshot,
			getProjectVersion: () => {
				return ctx.projectVersion.toString();
			},
			getProjectReferences: () => ctx.options.projectReferences,
			getCancellationToken: ctx.options.host!.getCancellationToken ? () => ctx.options.host!.getCancellationToken!() : undefined,
		};
		const vueTsLs = vueTs.createLanguageService(languageHost, vueCompilerOptions, ts as any, ts.sys);

		program = vueTs.getProgram(ts as any, vueTsLs.__internal__.context, vueTsLs, ts.sys) as (ts.Program & { __vue: ProgramContext; });
		program.__vue = ctx;

		function getVueCompilerOptions(): Partial<vue.VueCompilerOptions> {
			const tsConfig = ctx.options.options.configFilePath;
			if (typeof tsConfig === 'string') {
				return vue.createParsedCommandLine(ts as any, ts.sys, tsConfig).vueOptions;
			}
			return {};
		}
		function getScriptSnapshot(fileName: string) {
			return getScript(fileName)?.scriptSnapshot;
		}
		function getScript(fileName: string) {

			const script = scripts.get(fileName);
			if (script?.projectVersion === ctx.projectVersion) {
				return script;
			}

			const modifiedTime = ts.sys.getModifiedTime?.(fileName)?.valueOf() ?? 0;
			if (script?.modifiedTime === modifiedTime) {
				return script;
			}

			if (ctx.options.host!.fileExists(fileName)) {
				const fileContent = ctx.options.host!.readFile(fileName);
				if (fileContent !== undefined) {
					const script = {
						projectVersion: ctx.projectVersion,
						modifiedTime,
						scriptSnapshot: ts.ScriptSnapshot.fromString(fileContent),
						version: ctx.options.host!.createHash?.(fileContent) ?? fileContent,
					};
					scripts.set(fileName, script);
					return script;
				}
			}
		}
	}
	else {
		const ctx: ProgramContext = program.__vue;
		ctx.options = options;
		ctx.projectVersion++;
	}

	const vueCompilerOptions = program.__vue.vueCompilerOptions;
	if (vueCompilerOptions?.hooks) {
		const index = (state.hook?.index ?? -1) + 1;
		if (index < vueCompilerOptions.hooks.length) {
			const hookPath = vueCompilerOptions.hooks[index];
			const hook: Hook = require(hookPath);
			state.hook = {
				program,
				index,
				worker: (async () => await hook(program))(),
			};
			throw 'hook';
		}
	}

	for (const rootName of options.rootNames) {
		// register file watchers
		options.host.getSourceFile(rootName, ts.ScriptTarget.ESNext);
	}

	return program;
}

function toThrow(msg: string) {
	console.error(msg);
	return msg;
}
