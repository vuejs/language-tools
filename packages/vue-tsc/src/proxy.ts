import * as ts from 'typescript/lib/tsserverlibrary';
import * as vue from '@volar/vue-language-core';
import * as vueTs from '@volar/vue-typescript';

export function createProgramProxy(
	options: ts.CreateProgramOptions, // rootNamesOrOptions: readonly string[] | CreateProgramOptions,
	_options?: ts.CompilerOptions,
	_host?: ts.CompilerHost,
	_oldProgram?: ts.Program,
	_configFileParsingDiagnostics?: readonly ts.Diagnostic[],
) {

	if (!options.options.noEmit && !options.options.emitDeclarationOnly)
		return doThrow('js emit is not support');

	if (!options.host)
		return doThrow('!options.host');

	let program = options.oldProgram as any;

	if (!program) {

		const ctx = {
			projectVersion: 0,
			options: options,
		};
		const vueCompilerOptions = getVueCompilerOptions();
		const scripts = new Map<string, {
			projectVersion: number,
			modifiedTime: number,
			scriptSnapshot: ts.IScriptSnapshot,
			version: string,
		}>();
		const vueLsHost = new Proxy(<vue.LanguageServiceHost>{
			resolveModuleNames: undefined, // avoid failed with tsc built-in fileExists
			writeFile: (fileName, content) => {
				if (fileName.indexOf('__VLS_') === -1) {
					ctx.options.host!.writeFile(fileName, content, false);
				}
			},
			getCompilationSettings: () => ctx.options.options,
			getVueCompilationSettings: () => vueCompilerOptions,
			getScriptFileNames: () => {
				return ctx.options.rootNames as string[];
			},
			getScriptVersion,
			getScriptSnapshot,
			getProjectVersion: () => {
				return ctx.projectVersion.toString();
			},
			getProjectReferences: () => ctx.options.projectReferences,

			getTypeScriptModule: () => ts,
			isTsc: true,
		}, {
			get: (target, property) => {
				if (property in target) {
					return target[property as keyof vue.LanguageServiceHost];
				}
				return ctx.options.host![property as keyof ts.CompilerHost];
			},
		});
		const vueTsLs = vueTs.createLanguageService(vueLsHost);

		program = vueTsLs.getProgram();
		program.__VLS_ctx = ctx;

		function getVueCompilerOptions(): vue.VueCompilerOptions {
			const tsConfig = ctx.options.options.configFilePath;
			if (typeof tsConfig === 'string') {
				return vue.createParsedCommandLine(ts, ts.sys, tsConfig).vueOptions;
			}
			return {};
		}
		function getScriptVersion(fileName: string) {
			return getScript(fileName)?.version ?? '';
		}
		function getScriptSnapshot(fileName: string) {
			return getScript(fileName)?.scriptSnapshot;
		}
		function getScript(fileName: string) {

			const script = scripts.get(fileName);
			if (script?.projectVersion === ctx.projectVersion) {
				return script;
			}

			const modifiedTime = ts.sys.getModifiedTime?.(fileName)?.valueOf() ?? 0;
			if (script?.modifiedTime === modifiedTime) {
				return script;
			}

			if (ctx.options.host!.fileExists(fileName)) {
				const fileContent = ctx.options.host!.readFile(fileName);
				if (fileContent !== undefined) {
					const script = {
						projectVersion: ctx.projectVersion,
						modifiedTime,
						scriptSnapshot: ts.ScriptSnapshot.fromString(fileContent),
						version: ctx.options.host!.createHash?.(fileContent) ?? fileContent,
					};
					scripts.set(fileName, script);
					return script;
				}
			}
		}
	}
	else {
		program.__VLS_ctx.options = options;
		program.__VLS_ctx.projectVersion++;
	}

	for (const rootName of options.rootNames) {
		// register file watchers
		options.host.getSourceFile(rootName, ts.ScriptTarget.ESNext);
	}

	return program;
}

export function loadTsLib() {
	return ts;
}

function doThrow(msg: string) {
	console.error(msg);
	throw msg;
}
