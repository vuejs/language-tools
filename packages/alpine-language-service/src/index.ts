import * as alpineTs from '@volar/alpine-language-core';
import * as vueLs from '@volar/vue-language-service';
import { ConfigurationHost, EmbeddedLanguageServicePlugin, setCurrentConfigurationHost } from '@volar/vue-language-service-types';

export function createLanguageService(
	alpineLsHost: alpineTs.LanguageServiceHost,
	configurationHost: ConfigurationHost | undefined,
	customPlugins: EmbeddedLanguageServicePlugin[],
): ReturnType<typeof vueLs['createLanguageService']> {
	setCurrentConfigurationHost(configurationHost);
	return vueLs.createLanguageService(
		{
			...alpineLsHost,
			getVueCompilationSettings: () => ({}),
		},
		undefined,
		undefined,
		configurationHost,
		customPlugins,
		undefined,
		() => alpineTs.createLanguageContext(alpineLsHost),
	);
}

export function getDocumentService(
	ts: typeof import('typescript/lib/tsserverlibrary'),
	configurationHost: ConfigurationHost | undefined,
	customPlugins: EmbeddedLanguageServicePlugin[],
): ReturnType<typeof vueLs['getDocumentService']> {
	setCurrentConfigurationHost(configurationHost);
	return vueLs.getDocumentService(
		ts,
		configurationHost,
		undefined,
		customPlugins,
	);
}
