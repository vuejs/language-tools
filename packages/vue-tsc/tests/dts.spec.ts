import * as path from 'path';
import * as fs from 'fs';
import * as ts from 'typescript';
import { describe, expect, it } from 'vitest';
import { createProgram } from '../src';

const workspace = path.resolve(__dirname, '../../vue-test-workspace/vue-tsc-dts');
const testDirs = fs.readdirSync(workspace);
const ensureTs = (filename: string) => filename.endsWith('.ts') ? filename : filename + '.ts';
const normalizePath = (filename: string) => filename.replace(/\\/g, '/');

const compilerOptions: ts.CompilerOptions = {
	rootDir: workspace,
	declaration: true,
	emitDeclarationOnly: true,
};
const host = ts.createCompilerHost(compilerOptions);

describe('vue-tsc-dts', () => {
	for (const dirName of testDirs) {
		const dir = path.join(workspace, dirName);
		const files = fs.readdirSync(dir).map(file => path.join(dir, file));
		const program = createProgram({
			host,
			rootNames: files,
			options: compilerOptions
		});
		const service = program.__vue.languageService;
		for (const file of files) {
			const output = service.getEmitOutput(ensureTs(file), true);
			for (const outputFile of output.outputFiles) {
				it(`Input: ${normalizePath(path.join(dirName, path.basename(file)))}, Output: ${normalizePath(path.join(dirName, path.basename(outputFile.name)))}`, () => {
					expect(outputFile.text).toMatchSnapshot();
				});
			}
		}
	}
});
