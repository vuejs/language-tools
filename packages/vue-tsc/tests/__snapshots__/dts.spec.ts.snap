// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vue-tsc-dts > Input: basic/main.vue, Output: basic/main.vue.d.ts 1`] = `
"declare const _default: import(\\"vue\\").DefineComponent<{
    name: {
        type: import(\\"vue\\").PropType<string>;
        required: true;
    };
}, {}, unknown, {}, {}, import(\\"vue\\").ComponentOptionsMixin, import(\\"vue\\").ComponentOptionsMixin, {}, string, import(\\"vue\\").VNodeProps & import(\\"vue\\").AllowedComponentProps & import(\\"vue\\").ComponentCustomProps, Readonly<import(\\"vue\\").ExtractPropTypes<{
    name: {
        type: import(\\"vue\\").PropType<string>;
        required: true;
    };
}>>, {}, {}>;
export default _default;
"
`;
